pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,470 @@
+
+	if resolutions, exists := c.bindingNameToResolutions[bindingName]; exists {
+		_, exists = resolutions[objectIdentifier]
+		return exists
+	}
+
+	return false
+}
+
+// fetchMissingStatusCollectorSpecs fetches the missing statuscollector specs
+// from the given lister and updates the cache.
+// The method is expected to be called with the write lock held.
+func (c *combinedStatusResolver) fetchMissingStatusCollectorSpecsLocked(statusCollectorLister controllisters.StatusCollectorLister,
+	statusCollectorNames []string) {
+	for _, statusCollectorName := range statusCollectorNames {
+		if _, exists := c.statusCollectorNameToSpec[statusCollectorName]; exists {
+			continue // this method is not responsible for keeping the cache up-to-date
+		}
+
+		statusCollector, err := statusCollectorLister.Get(statusCollectorName)
+		if err != nil {
+			// fetch error should not disturb the flow.
+			// a missing spec will be reconciled when the status collector is created/updated.
+			runtime2.HandleError(fmt.Errorf("failed to get statuscollector %s: %w", statusCollectorName, err))
+		}
+
+		c.statusCollectorNameToSpec[statusCollectorName] = &statusCollector.Spec // readonly
+	}
+}
+
+// evaluateWorkStatusesPerBindingLocked evaluates workstatuses associated with the given
+// workload identifiers and destinations. Workstatuses that are not associated
+// with the two are skipped.
+// The returned set contains the identifiers of combinedstatus objects that
+// should be queued for syncing.
+// The method is expected to be called with the read lock held.
+func (c *combinedStatusResolver) evaluateWorkStatusesPerBindingLocked(bindingName string,
+	workloadIdentifiersToEvaluate sets.Set[util.ObjectIdentifier], destinations sets.Set[string],
+	workStatuses []*workStatus) sets.Set[util.ObjectIdentifier] {
+	workloadObjectToError := make(map[util.ObjectIdentifier]error) // TODO: use
+	combinedStatusesToQueue := sets.Set[util.ObjectIdentifier]{}
+
+	for _, workStatus := range workStatuses {
+		// skip workstatuses that are not associated with the binding's destinations
+		if destinations.Has(workStatus.wecName) {
+			continue
+		}
+
+		// skip workstatuses that are not associated with the workload objects
+		if !workloadIdentifiersToEvaluate.Has(workStatus.sourceObjectIdentifier) {
+			continue
+		}
+
+		// evaluate workstatus
+		csResolution := c.bindingNameToResolutions[bindingName][workStatus.sourceObjectIdentifier]
+		changed, err := csResolution.evaluateWorkStatus(c.celEvaluator, workStatus.wecName, workStatus.status)
+		if err != nil {
+			workloadObjectToError[workStatus.sourceObjectIdentifier] = err
+		}
+
+		if changed {
+			combinedStatusesToQueue.Insert(getCombinedStatusIdentifier(bindingName, workStatus.sourceObjectIdentifier))
+		}
+	}
+
+	return combinedStatusesToQueue
+}
+
+func statusCollectorSpecsMatch(spec1, spec2 *v1alpha1.StatusCollectorSpec) bool {
+	if spec1.Limit != spec2.Limit {
+		return false
+	}
+
+	if spec1.Filter != spec2.Filter {
```

Comment:
This is not the right way to compare pointers to strings.

Reply:
I think that this func needs to compare the referenced strings, not the pointers to the strings.